src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,974 @@
+                var requiredArguments = new List<SyntaxNode>();
+
+                foreach (var parameterSymbol in validParameters)
+                {
+                    if (parameterToArgumentMap.TryGetValue(parameterSymbol, out var index))
+                    {
+                        requiredArguments.Add(currentInvocationArguments[index]);
+                    }
+                }
+
+                var conditionalRoot = syntaxFacts.GetRootConditionalAccessExpression(invocation);
+                var named = ShouldArgumentBeNamed(compilation, semanticFacts, invocationSemanticModel, invocationArguments, insertionIndex, cancellationToken);
+                var newMethodInvocation = GenerateNewMethodInvocation(syntaxFacts, generator, invocation, requiredArguments, newMethodIdentifier);
+
+                SeparatedSyntaxList<SyntaxNode> allArguments;
+                if (conditionalRoot is null)
+                {
+                    allArguments = AddArgumentToArgumentList(currentInvocationArguments, generator, newMethodInvocation, parameterName, insertionIndex, named);
+                }
+                else
+                {
+                    var expressionsWithConditionalAccessors = conditionalRoot.ReplaceNode(invocation, newMethodInvocation);
+                    allArguments = AddArgumentToArgumentList(currentInvocationArguments, generator, expressionsWithConditionalAccessors, parameterName, insertionIndex, named);
+                }
+
+                return UpdateArgumentListSyntax(currentArgumentListSyntax, allArguments);
+            }
+        }
+
+        private static SyntaxNode GenerateNewMethodInvocation(ISyntaxFactsService syntaxFacts, SyntaxGenerator generator, SyntaxNode invocation, List<SyntaxNode> arguments, string newMethodIdentifier)
+        {
+            var methodName = generator.IdentifierName(newMethodIdentifier);
+            var fullExpression = syntaxFacts.GetExpressionOfInvocationExpression(invocation);
+            if (syntaxFacts.IsAnyMemberAccessExpression(fullExpression))
+            {
+                var identifierExpression = syntaxFacts.GetExpressionOfMemberAccessExpression(fullExpression);
+                if (identifierExpression is not null)
+                {
+                    methodName = generator.MemberAccessExpression(identifierExpression, newMethodIdentifier);
+                }
+                else
+                {
+                    // In the VB case sometimes the the member access expression's expression is nothing
+                    // so I generate a member binding expression instead.
+                    methodName = generator.MemberAccessExpression(expression: null, newMethodIdentifier);
+                }
+            }
+            else if (syntaxFacts.IsMemberBindingExpression(fullExpression))
+            {
+                methodName = generator.MemberBindingExpression(generator.IdentifierName(newMethodIdentifier));
```

Comment:
do you have tests where the original method is generic?  is the presumption that we don't have to pass type arguments along because they'll all be inferred?

Reply:
if so, that seems olike an overly aggressive assumption.  however, determining if we need type args or not is challenging, so we can leave out os cope for now.

Reply:
I do not have tests for that, I can add it